### ISSUE TITLE
Failing to pull certain metrics on Slurm 24.05

### ISSUE THREAD

**Issue by thsu-terraytx** [lcrownover/prometheus-slurm-exporter#21]
Hello, 

I'm running the latest version with Slurm 24.05 and noticed certain metrics were not pulling to prometheus. 

These are the metrics I couldn't see:
- partitions 
- account
- cpu
- queue

I looked at the exporter service and it looks like there were relating error messages for these metrics- see below.

I also confirmed if I manually queried the Slurm API using the JWT token that it returns the results fine for diag, jobs, shares, partitions, etc. 
__________
- level=ERROR msg="failed to unmarshal partitions response for partitions metrics" error="failed to unmarshall partitions response data: unexpected end of JSON input"
- level=ERROR msg="failed to unmarshal jobs response for accounts metrics" error="failed to unmarshall job response data: unexpected end of JSON input"
- level=ERROR msg="failed to unmarshal jobs response for cpu metrics" error="failed to unmarshall job response data: unexpected end of JSON input"
- level=ERROR msg="failed to unmarshal jobs response for queue metrics" error="failed to unmarshall job response data: unexpected end of JSON input"
- level=ERROR msg="error populating request cache" error="errors encountered calling slurm api: failed to get slurmrestd partitions response: received incorrect status code for partition data"

**Comment by thsu-terraytx** [comment]
You can disregard the error above, it was related to a config issue with the API but I resolved it. 

However, I'm still not able to see the 4 metrics above.

**Comment by lcrownover** [comment]
Hey @thsu-terraytx, thanks for the report! Would you mind running the exporter with `SLURM_EXPORTER_DEBUG=1` and posting the relevant debug information near these errors?

**Caution:** The debug option might display some sensitive items so please sanitize the data before posting

**Comment by thsu-terraytx** [comment]
@lcrownover I turned on debug mode but didn't see anymore errors appear.
However, when I look at the metric in Prometheus it does look like its generating data but only when jobs are being run. 

It looks like if there is nothing running it doesn't output or gather any data, not even a 0.
See example of a chart below where you see metric for "pending jobs per partition" but after all jobs completed it sends no data. 

<img width="414" alt="image" src="https://github.com/user-attachments/assets/a2af3814-5411-4bc8-88ca-96f5f2732992">

**Comment by lcrownover** [comment]
@thsu-terraytx Thanks for that information, I have a feeling that the blanks in the chart are caused by the errors above, where it seems like it's unable to pull/unmarshal the required data.

That last error message is interesting, and I can trace it back to this conditional:

```go
if resp.StatusCode != 200 {
    slog.Debug("incorrect status code for partition data", "code", resp.StatusCode, "body", string(resp.Body))
    return nil, fmt.Errorf("received incorrect status code for partition data")
}
```

While you're correct that the other unmarshal errors don't show the response, this one should have dumped the response status code and body to debug output. When you have debug on, do you see the initial "debug logging enabled" string printed just before the "Starting Prometheus Exporter version..." ? I'd be curious to know what that debug error looks like.

**Comment by thsu-terraytx** [comment]
@lcrownover Yeah- I do see the service running with DEBUG logging enabled. But I just haven't seen any more of the pull/marshal errors appearing since enabling.

**Comment by lcrownover** [comment]
@thsu-terraytx Gotcha. I do see that it's happening intermittently in my logs as well so I'll do some debugging and let you know when it's fixed!

**Comment by thsu-terraytx** [comment]
@lcrownover Here are the DEBUG error logs if you still need it.

- level=ERROR msg="failed to unmarshal jobs response for queue metrics" error="failed to unmarshall job response data: no value given for required property select_type"
- level=ERROR msg="failed to unmarshal jobs response for users metrics" error="failed to unmarshall job response data: no value given for required property select_type"
- level=ERROR msg="failed to unmarshal jobs response for cpu metrics" error="failed to unmarshall job response data: no value given for required property select_type"
- level=ERROR msg="failed to unmarshal jobs response for partitions metrics" error="failed to unmarshall job response data: no value given for required property select_type"
- level=ERROR msg="failed to unmarshal jobs response for accounts metrics" error="failed to unmarshall job response data: no value given for required property select_type"

**Comment by lcrownover** [comment]
That's super helpful, thank you!

**Comment by lcrownover** [comment]
I've got a large update for the exporter nearly ready that fixes several bugs and includes better logging for errors like this. I just need to do a little more testing before I create a release.

It's possible that you (like me) are running into actual slurm errors, which I've created a schedmd ticket for here: https://support.schedmd.com/show_bug.cgi?id=21253

**If** that's the case, which we'll be able to confirm in the next release, then there's nothing I can do to resolve those errors as they exist in slurm.

I'll keep you posted!

**Comment by rohithmn01** [comment]
@lcrownover 

As the schedmd ticket states RESOLVED - When are you releasing the new update. Waiting for the next release.!

**Comment by lcrownover** [comment]
@rohithmn01 

We're just finishing up our integration test suite here hopefully next week so testing multiple version of the exporter can be done a bit easier. 

Thanks for your patience, lots going on this time of year 😊

**Comment by rohithmn01** [comment]
@lcrownover 

Thanks for the update. I am not sure, whether you have already considered this. Even `nodes` data it couldn't be able to pull.
**Debug logs:**

```
level=DEBUG msg="debug logging enabled"
level=INFO msg="Starting Prometheus Slurm Exporter 1.0.11"
level=INFO msg="Starting Server: 0.0.0.0:8080"
level=ERROR msg="failed to collect cpus metrics" error="failed to get node state for cpu metrics: failed to match cpu state against known states: dynamic_norm"
level=ERROR msg="failed to collect nodes metrics" error="failed to get node state for nodes metrics: failed to match cpu state against known states: dynamic_norm"
level=ERROR msg="failed to collect nodes metrics" error="failed to get node state: failed to get node states: failed to match cpu state against known states: dynamic_norm"
```
As stated above the Slurm API using the JWT token is working fine.!

**Comment by lcrownover** [comment]
Thanks for that report, should be fixed in the next release: https://github.com/lcrownover/prometheus-slurm-exporter/commit/a41ac468eec8b9ee8be9269fb21495d05638dd31

**Comment by rohithmn01** [comment]
@lcrownover 

Just checking back on the new release.! Any expected date :)

**Comment by lcrownover** [comment]
Hey @rohithmn01 @thsu-terraytx , I've released a new version 2.0.0-beta: https://github.com/lcrownover/prometheus-slurm-exporter/releases/tag/v2.0.0-beta

Per the commit message for that merge, we still don't have a robust integrated testing environment set up because it's not trivial to simulate a production slurm cluster (with nodes in various states, jobs by different users, etc). 

This release includes several bug fixes, but also involves a large refactor of the codebase to minimize build tags, which should reduce friction in development.

That being said, I don't think there should be any issue with this release, but I've marked it as beta, just in case there's a breaking change that I'm not expecting. We're running this beta version at UO, any it's been working just fine.

Please do let me know if there are any bugs you find and I can patch them asap!

**Comment by thsu-terraytx** [comment]
Great- thanks @lcrownover I'll give that version a try.

**Comment by thsu-terraytx** [comment]
@lcrownover Just an update, so far everything looks good on the beta version, no errors and no weird gaps in the metric data.

**Comment by lcrownover** [comment]
Thanks for testing @thsu-terraytx ! I've made a new release and not marked it beta.
